### ISSUE TITLE
Comparison Series Search bar issue

### ISSUE THREAD

**Issue by workingbuddy10** [amcharts/amcharts5#1240]
I am using comparison series in my stockchart, I am using the inbuilt search bar of comparison series. 

I have more than 2000 products from where I have to search. But when I try to search any product my whole app gets stuck/freezes. I believe there is some issue in handling large number of search products in the search bar.

You can try to search `zaa ` product name in the search bar in the below link:
https://codepen.io/Ansh-m-the-reactor/pen/BabBayx

**Comment by martynasma** [comment]
I strongly suggest populating your list using search function instead:
https://www.amcharts.com/docs/v5/charts/stock/toolbar/comparison-control/#Dynamic_search_results
https://www.amcharts.com/docs/v5/charts/stock/toolbar/dropdown-list-control/#Dynamic_search_results

Manipulating so many DOM items is not efficient.

**Comment by workingbuddy10** [comment]
In my app code I have done/added like this, what changes should I add :
```
    //comparison series
    var comparisonControl = am5stock.ComparisonControl.new(root, {
      stockChart: stockChart,
      name: "",
      description: "Comparison Control",
      searchable: true,
      items: props.products,
    });

```

@martynasma

**Comment by martynasma** [comment]
Here's a demo that uses `searchCallback` for comparison dropdown:
https://www.amcharts.com/demos/stock-chart-comparing-prices/

**Comment by workingbuddy10** [comment]
@martynasma Did the same as you suggested!

In the case where user selectes more than 4 products: Though it shows error/notify the user when they SEARCH the product from the search bar. But it doesn't notify when user is not searching just clicking it from the options.

I am showing like this, so when I don't search it from the search bar and just plot products it allows me to plot more than4 products but when I search it from the search bar then only it notifies me.
![image](https://github.com/amcharts/amcharts5/assets/137867461/5b28d246-f881-4ff1-8a6e-dfb9b6356556)

**Comment by martynasma** [comment]
I don't think mixing preset list of items and searchCallback is a good idea for big lists.

I suggest you leave only search. Not that someone would scroll a list of 2000 items looking for something in it.

**Comment by workingbuddy10** [comment]
Very True @martynasma 

But I am planning to give a feature where user can mark the favorite products, so from the drop-down user can mark there products to favorite so that they don't have to search again and again, as favorite marked products comes at the top. 

I guess this is better approach!

So in this case, user don't have to search the products and he will get there favorite products at the top only, so now he just have to click on the products without searching.

Though there can also be a case where sometimes user can search the products, so we to consider this case also.

**Comment by martynasma** [comment]
> But I am planning to give a feature where user can mark the favorite products, so from the drop-down user can mark there products to favorite so that they don't have to search again and again, as favorite marked products comes at the top.

I'm guessing there won't be 2000 "favorites". Just a few, which makes sense to keep them displayed automatically, but leave search functionality for the complete list.

**Comment by workingbuddy10** [comment]
How can I change the color of the label/text as in the dark theme it's not visible that much.

```
if (compared.length > 4) {
          return [
            {
              label: "A maximum of 5 comparisons added",
              subLabel: "Remove some to add new ones",
              id: "",
              className: "am5stock-list-info",
            },
          ];
        }
```
![image](https://github.com/amcharts/amcharts5/assets/137867461/67ac9664-6353-42ae-ab87-9cf0e664700d)

@martynasma

**Comment by workingbuddy10** [comment]
> Here's a demo that uses `searchCallback` for comparison dropdown: https://www.amcharts.com/demos/stock-chart-comparing-prices/

In this demo only, Even though I removed one product then to it's showing me to remove the products.

Though I know when I'll search then it will be removed, but it is giving the false information to the USERS even though they have removed the product.

https://www.loom.com/share/cb97119b7106489f8a6a4bdfbc713da8?sid=5d071291-35c8-4669-aba1-59bc29539386

@martynasma

**Comment by workingbuddy10** [comment]
And second issue Even though there is no data for testing product then too :
- It placed the legend of testing.
- Updated the Y-axis with percent values.

Ideally there should be no changes in the Y-axis values & there should be no legend as there is no data in the testing product.

![image](https://github.com/amcharts/amcharts5/assets/137867461/dcc4a182-388c-4095-82b0-938c19468f4f)

@martynasma

**Comment by martynasma** [comment]
I'm sorry, I don't understand what you're saying.

I can't reproduce what you described or shown in the video on this demo:
https://www.amcharts.com/demos/stock-chart-comparing-prices/

**Comment by workingbuddy10** [comment]
> I'm sorry, I don't understand what you're saying.
> 
> I can't reproduce what you described or shown in the video on this demo: https://www.amcharts.com/demos/stock-chart-comparing-prices/

1. Firstly I added 5 products in the comparison series. 
2. And then I searched for one more product, it showed that I have to remove any product series. Till here everything is fine.
3. Then I removed a comparison series product (TESLA) and then again I opened (I didn't searched anything just opened) the comparison series option again then too it is showing me to remove  a product.

https://www.loom.com/share/7c57a625b4c2433ca0a640df102b0df2?sid=92e49bfb-96b2-41e1-a9cb-23090905d829

Ideally what should have happened at step 3 that it should not show the message to remove the product as there are only 4 products plotted


https://codepen.io/Ansh-m-the-reactor/pen/oNVvEBK

@martynasma

**Comment by workingbuddy10** [comment]
> And second issue Even though there is no data for testing product then too :
> 
> * It placed the legend of testing.
> * Updated the Y-axis with percent values.
> 
> Ideally there should be no changes in the Y-axis values & there should be no legend as there is no data in the testing product.
> 
> ![image](https://private-user-images.githubusercontent.com/137867461/291870449-dcc4a182-388c-4095-82b0-938c19468f4f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNzY0MzMsIm5iZiI6MTcwMzA3NjEzMywicGF0aCI6Ii8xMzc4Njc0NjEvMjkxODcwNDQ5LWRjYzRhMTgyLTM4OGMtNDA5NS04MmIwLTkzOGMxOTQ2OGY0Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyMFQxMjQyMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMjczODc5NzM1OTRmZTdjZDQ3NzBlMTUxNGE0Mjk0NjE2M2YyZTIwMmEwOTM5NmMyZTMzOWU4NDc1ODQ3YmUwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9kKbvP5gpuelWgngqIqrMpNabNkZacHh8E3RMMkeki0)
> 
> @martynasma

And what about this?

**Comment by workingbuddy10** [comment]
One more bug in the comparison series search:

Even it is adding the text as a series " Remove some to add new ones" I guess it is happening because of label property given.

https://www.loom.com/share/21d314bfc3af4213a00d3957af1f4c9f?sid=595e4d6d-f15e-4ed9-954e-2c4a39ced445

**Comment by martynasma** [comment]
Yeah, that is a bug. Thanks for the explanation.

We'll fix it in the next version.

**Comment by martynasma** [comment]
OK, so the issue was with the demo, not the StockChart.

Demo code updated.

**Comment by workingbuddy10** [comment]
@martynasma Still there are 2 issues in this :

Issue 1 :
Even after removing the 5th comparison series from the chart, yet when I open the comparison tab it shows the message to remove the product.

Ideally when I removed the 5th product then I should not get the message to remove the product.

How to reproduce this: Just try to add the 6th product it will show the message & then remove the entered letter. Then remove one series and again open the comparison menu.

Issue 2:
When I don't have any values in the product or that product doesn't exist then too it is placing the legend and updating the Y-axis.
Ideally if the product doesn't have data then it shouldn't add & update the legend of that product & Y-axis.

Video to view both the issues respectively:

https://www.loom.com/share/dcdd8c4dbafc45629624a582378fd72b?sid=c183b0b5-795f-46a3-aab3-b4727445f2a2

**Comment by martynasma** [comment]
> Even after removing the 5th comparison series from the chart, yet when I open the comparison tab it shows the message to remove the product.

This will be fixed in the next version. Thanks.

> When I don't have any values in the product or that product doesn't exist then too it is placing the legend and updating the Y-axis.
Ideally if the product doesn't have data then it shouldn't add & update the legend of that product & Y-axis.

The series and legend is added by a custom code which you control. Therefore I think it's appropriate for you to augment your code to handle situations where there is no data, so that such series are not added.

Ideally, data-less items should not make into search list at all. But that's up to you to implement 😉

**Comment by martynasma** [comment]
Fixed in [5.7.2](https://github.com/amcharts/amcharts5/releases/tag/5.7.2).

## [5.7.2] - 2023-12-22

### Added
- New setting on `Label`: `maxChars`. If set to a number, will truncate text to X characters with ellipsis, obeying `breakWords` and `ellipsis` settings.
- "Bull Bear Power" indicator added to `StockChart`.
- "Acceleration Bands" indicator added to `StockChart`.
- Standalone `registerClass(name, classRef)` added to Stock Chart package (global: `am5stock.registerClass()`). Allows registering own indicator class so that it gets properly serialized and restored.

### Changed
- Improved min/max calculation of the `ValueAxis` when all the values of a series are the same.

### Fixed
- "Volume Profile" was showing up in the `IndicatorControl`'s dropdown, even if `volumeSeries` was not set on the `StockChart`.
- Left-side `ValueAxis` on a `StockChart` (`opposite: false`) it was impossible to right-align axis labels to plot area. Scrollbar's width was also not adjusted to the width of the plot area.
- Some drawing tools in `StockChart` where not working when chart was in comparison mode.
- Drawings created when `StockChart` was in comparinson mode were not carried over correctly when switching to regular value mode.
- Some drawing tools ceized working when main `StockChart` series was updated.
- Searchable dropdown controls in `StockChart` were not clearing up the list of found items when the search field was cleared.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
